docs: gatsby for apps - auth and client only routes (#19551)
gatsbyjs/gatsby: docs/docs/building-a-site-with-authentication.md

title: Building a Site with Authentication
---

Many sites require users to be authenticated in order to protect private data.

## Understanding authentication between client and server

In many modern websites, the [client](/docs/glossary#client-side) -- or [frontend](/docs/glossary#frontend) -- is [decoupled](/docs/glossary#decoupled) from the [backend](/docs/glossary#backend). This pattern is how Gatsby functions to combine data from a myriad of backend sources to facilitate building the frontend.

> In order to provide authentication functionality, another service has to be leveraged and connected to Gatsby. Examples of these kinds of services could be open source technology like:

Reviewer: ```suggestion
In order to provide authentication functionality, another service has to be leveraged and connected to Gatsby. There are many open source tools than can provide this functionality. Examples inclue:
```

- A Node.js app using Passport.js
- A Ruby on Rails API using Devise

> Or from a third party like:

Reviewer: ```suggestion
Another option are third party technologies like:
```

- Firebase
- Auth0
- AWS Amplify
- Netlify Identity

> The process of verifying a user on the client against an authentication service and using a token returned by it to get protected data is similar for any of these services and is visualized in this diagram:

Reviewer: ```suggestion
These tools follow a process in order to verify a user on the client against an authentication service. The service returns a token that the client can use to access protected data. This diagram visualizes the process:
```

![Diagram of Gatsby using an authentication service to get data from an API](./images/basic-auth.png)

1. First, a request is made from a Gatsby site (the client) to an authentication service to perform an action like register a new user or login.

> 2. If the credentials (like a username and password) provided from the client match a user in the authentication service, it returns a token (like a JSON Web Token, abbreviated as JWT) so that the user has a key that they can use to prove they are who they say they are. The data returned about the user can be stored in the Gatsby app and passed to components using a Provider component with the React Context API using the [`wrapRootElement` API](/docs/browser-apis/#wrapRootElement) from Gatsby.

Reviewer: ```suggestion
2. If the credentials (like a username and password) provided from the client match a user in the authentication service, it returns a token (like a JSON Web Token, abbreviated as JWT) so the user has a key they can use to prove they are who they say they are. The user data returned can be stored in the Gatsby app by passing it to components using a Provider component with the React Context API and [`wrapRootElement` API](/docs/browser-apis/#wrapRootElement) from Gatsby.
```

> 3. With the key, the client can make a request to an external data source like an API (the server) where protected data is stored, the key is unique to a specific user that allows them to access their specific data.

Reviewer: ```suggestion
3. With the key, the client can make a request to an external data source like an API (the server) where protected data is stored. The key is unique to a specific user and allows the client to access their specific data.
```

> 4. The server returns data back to the client that it can use on the client to pass information into components.

Reviewer: ```suggestion
4. The server returns data back to the client that it can use to pass information into components.
```

_**Note**: this is the same pattern that other sites built with React (like Create React App) would need to follow._

## Implementing authentication in a Gatsby site

There are a few things to be aware of when implementing authentication in a Gatsby site, because of how Gatsby uniquely builds pages and renders static assets with dynamic capabilities.

### Setting up client-only routes

With Gatsby, you are able to create restricted areas in your app using [client-only routes](/docs/building-apps-with-gatsby/#client-only-routes).

> Because static HTML files generated by Gatsby sit on a file server, you cannot programmatically control access to those files on the server from Gatsby (for example: a user could guess or type in a URL and navigate straight to the page). As the [section from the Adding App and Website functionality](/docs/adding-app-and-website-functionality/#client-only-routes) overview page demonstrates, client-only routes can be created to route a user between pages using a React-based router, as opposed to navigating between different static HTML files on a server. Gatsby is different from a traditional Single Page Application in

Reviewer 2: "in ..." I miss an end here

> Because static HTML files generated by Gatsby sit on a file server, you cannot programmatically control access to those files on the server from Gatsby (for example: a user could guess or type in a URL and navigate straight to the page). As the [section from the Adding App and Website functionality](/docs/adding-app-and-website-functionality/#client-only-routes) overview page demonstrates, client-only routes can be created to route a user between pages using a React-based router, as opposed to navigating between different static HTML files on a server. Gatsby is different from a traditional Single Page Application in

Reviewer: ```suggestion
Because static HTML files generated by Gatsby sit on a file server, you cannot programmatically control access to those files (for example: a user could guess or type in a URL and navigate straight to the page). As the [section from the Adding App and Website functionality](/docs/adding-app-and-website-functionality/#client-only-routes) overview page demonstrates, client-only routes can be created to route a user between pages using a React-based router, as opposed to navigating between different static HTML files on a server. Gatsby is different from a traditional Single Page Application in
```

Reviewer: Think something is missing at the end of this paragraph?

Author: Updated this and added a little section back in the overview about differences between Gatsby and "traditional" React apps.

> Taking advantage of this client-side routing gives the control back to you to protect or customize your routes. Using the [`@reach/router` library](https://reach.tech/router/), which comes installed with Gatsby, you can setup a router on a page and control which component will be loaded when a certain route is called, and check for the existence of a variable like authentication state before serving the content.

Reviewer: ```suggestion
Taking advantage of this client-side routing allows you to protect or customize your routes. Using the [`@reach/router` library](https://reach.tech/router/), which comes installed with Gatsby, you can set up a router on a page and control which component loads when a certain route is called, and check for the existence of a variable like authentication state before serving the content.
```

<!-- prettier-ignore -->
```jsx
<Router>
  {isAuthenticated ? <PrivateRoute /> : <Login />}
</Router>
```

More specific code examples for this pattern are outlined in the [Client-only Routes & User Authentication guide](/docs/client-only-routes-and-user-authentication/#implementing-client-only-routes).

### Protecting code from accessing browser globals during build

Global objects that are accessible in the browser like `localStorage` aren't available while a Gatsby site is building, because the build runs in a Node.js environment.

> Some third party services might try and access `localStorage` or the `window` object with internal methods, to keep those snippets from breaking the build, those invocations should be wrapped in checks or `useEffect` hooks to verify that the code is running in the browser and is skipped during the build process:

Reviewer: ```suggestion
However, some third party services might try and access `localStorage` or the `window` object with internal methods. To keep those snippets from breaking the build, those invocations should be wrapped in checks or `useEffect` hooks to verify that the code is running in the browser and is skipped during the build process:
```

```javascript
import app from "firebase/app"

...

if (typeof window !== 'undefined') { // highlight-line
  app.initializeApp(config)
} // highlight-line
```

More information on build related errors is available in the guide on [debugging HTML builds](/docs/debugging-html-builds/).

## Real-world example: Gatsby store

The [Gatsby store](https://github.com/gatsbyjs/store.gatsbyjs.org) is a live application built with Gatsby that implements authentication using Auth0.

> [Util functions](https://github.com/gatsbyjs/store.gatsbyjs.org/blob/master/src/utils/auth.js) in the repo make use of Auth0's APIs while checking that [some code only runs in the browser](https://github.com/gatsbyjs/store.gatsbyjs.org/blob/master/src/utils/auth.js#L3).

Reviewer: A bit confused by this sentence. Are the util functions making sure the code is wrapped properly? Or is it Auth0?

Author: Just made some updates, does it make more sense now?

> [Util functions](https://github.com/gatsbyjs/store.gatsbyjs.org/blob/master/src/utils/auth.js) in the Gatsby Store repo make use of Auth0's APIs to authenticate users with GitHub, and wrap Auth0's APIs to check that [some of the Auth0 code only runs in the browser](https://github.com/gatsbyjs/store.gatsbyjs.org/blob/master/src/utils/auth.js#L3).

Reviewer: ```suggestion
[Util functions](https://github.com/gatsbyjs/store.gatsbyjs.org/blob/master/src/utils/auth.js) in the Gatsby Store repo make use of Auth0's APIs to authenticate users with GitHub, and wrap Auth0's APIs to check that [some of the Auth0 code runs only in the browser](https://github.com/gatsbyjs/store.gatsbyjs.org/blob/master/src/utils/auth.js#L3).
```
